### PR TITLE
Fix race condition in DaemonSet controller. Fixes #14693.

### LINKED ISSUE #14693: flaky e2e: DaemonRestart Controller Manager should not create/delete replicas across restart
DaemonRestart Controller Manager should not create/delete replicas across restart (Failed 7 times in the last 30 runs. Stability: 76 %)

### PR DESCRIPTION
**Pull request by davidopp**
@bprashanth

### CODE REVIEW

**Comment by davidopp** [comment]
I ran the test about 20 times and didn't get any failures.

File: pkg/controller/daemon/controller.go [hunk 1]
```diff
@@ -155,6 +166,7 @@ func NewDaemonSetsController(kubeClient client.Interface) *DaemonSetsController
 		},
 	)
 	dsc.syncHandler = dsc.syncDaemonSet
+	dsc.podStoreSynced = dsc.podController.HasSynced
```
Comment (bprashanth) [inline pkg/controller/daemon/controller.go]: The reason I did this was so unittesting is easier. Please either add a similar unittest, or just invoke the function directly

Comment (davidopp) [inline pkg/controller/daemon/controller.go]: Added a similar unittest.

**Comment by davidopp** [comment]
PTAL.

**Comment by bprashanth** [comment]
LGTM

**Pull request merged**